File: day16.nim
import strutils, tables, re

const
  instructions = readFile("./inputs/16.txt").splitLines
  ticker = {
    "children": 3,
    "cats": 7,
    "samoyeds": 2,
    "pomeranians": 3,
    "akitas": 0,
    "vizslas": 0,
    "goldfish": 5,
    "trees": 3,
    "cars": 2,
    "perfumes": 1
  }.toTable

let pattern = re"(\w+):\s(\d+)"

func check(name: string, value: int): bool =
  case name
  of "cats", "trees":
    return ticker[name] < value
  of "pomeranians", "goldfish":
    return ticker[name] > value
  else:
    return ticker[name] == value


for i, line in instructions:
  var
    sue1 = true
    sue2 = true
  for m in line.findAll(pattern):
    let
      s = m.split(": ")
      name = s[0]
      value = s[1].parseInt
    if ticker[name] != value: sue1 = false
    if not check(name, value): sue2 = false
  if sue1: echo "first: ", i+1
  if sue2: echo "second: ", i+1

File: python/day16.py
import re


pattern = re.compile(r"(\w+): (\d+)")

ticker = {
    "children": 3,
    "cats": 7,
    "samoyeds": 2,
    "pomeranians": 3,
    "akitas": 0,
    "vizslas": 0,
    "goldfish": 5,
    "trees": 3,
    "cars": 2,
    "perfumes": 1
}

parse_line = lambda line: [(n, int(v)) for n, v in re.findall(pattern, line)]

check_1 = lambda name, val: ticker[name] == val
check_2 = lambda name, val: (
    val > ticker[name] if name in ("cats", "trees") else
    val < ticker[name] if name in ("pomeranians", "goldfish") else
    val == ticker[name])

solve = lambda data, func: next(i for i, line in enumerate(data, 1)
                                if all(func(*p) for p in line))


data = list(map(parse_line, open("inputs/16.txt").read().splitlines()))

print(solve(data, check_1))
print(solve(data, check_2))